src/messaging/ExchangeMgr.cpp

```diff
@@ -0,0 +1,344 @@
+    return ec;
+}
+
+ExchangeContext * ExchangeManager::FindContext(uint64_t peerNodeId, void * appState, bool isInitiator)
+{
+    ExchangeContext * ec = ContextPool;
+
+    for (int i = 0; i < CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS; i++, ec++)
+    {
+        if (ec->GetExchangeMgr() != nullptr && ec->GetPeerNodeId() == peerNodeId && ec->GetAppState() == appState &&
+            ec->IsInitiator() == isInitiator)
+            return ec;
+    }
+
+    return nullptr;
+}
+
+CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeContext::MessageReceiveFunct handler,
+                                                              void * appState)
+{
+    return RegisterUMH(protocolId, kAnyMessageType, handler, appState);
+}
+
+CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandler(uint32_t protocolId, uint8_t msgType,
+                                                              ExchangeContext::MessageReceiveFunct handler, void * appState)
+{
+    return RegisterUMH(protocolId, static_cast<int16_t>(msgType), handler, appState);
+}
+
+CHIP_ERROR ExchangeManager::UnregisterUnsolicitedMessageHandler(uint32_t protocolId)
+{
+    return UnregisterUMH(protocolId, kAnyMessageType);
+}
+
+CHIP_ERROR ExchangeManager::UnregisterUnsolicitedMessageHandler(uint32_t protocolId, uint8_t msgType)
+{
+    return UnregisterUMH(protocolId, static_cast<int16_t>(msgType));
+}
+
+void ExchangeManager::OnReceiveError(CHIP_ERROR error, const Transport::PeerAddress & source, SecureSessionMgrBase * msgLayer)
+{
+    ChipLogError(ExchangeManager, "Accept FAILED, err = %s", ErrorStr(error));
+}
+
+ExchangeContext * ExchangeManager::AllocContext()
+{
+    ExchangeContext * ec = ContextPool;
+
+    CHIP_FAULT_INJECT(FaultInjection::kFault_AllocExchangeContext, return nullptr);
+
+    for (int i = 0; i < CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS; i++, ec++)
+    {
+        if (ec->GetExchangeMgr() == nullptr)
+        {
+            *ec = ExchangeContext();
+            ec->SetExchangeMgr(this);
+            ec->SetRefCount(1);
+            mContextsInUse++;
+
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+            ChipLogProgress(ExchangeManager, "ec++ id: %d, inUse: %d, addr: 0x%x", (ec - ContextPool + 1), mContextsInUse, ec);
+#endif
+            SYSTEM_STATS_INCREMENT(chip::System::Stats::kExchangeMgr_NumContexts);
+
+            return ec;
+        }
+    }
+
+    ChipLogError(ExchangeManager, "Alloc ctxt FAILED");
+    return nullptr;
+}
+
+void ExchangeManager::DispatchMessage(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, PacketBuffer * msgBuf)
+{
+    UnsolicitedMessageHandler * umh         = nullptr;
+    UnsolicitedMessageHandler * matchingUMH = nullptr;
+    ExchangeContext * ec                    = nullptr;
+    CHIP_ERROR err                          = CHIP_NO_ERROR;
+
+    // Search for an existing exchange that the message applies to. If a match is found...
+    ec = (ExchangeContext *) ContextPool;
```

Comment:
could we do a cpp style cast here?

In other places no cast is needed though ... why is one required here?

Reply:
Ack